Userland/Libraries/LibJS/Runtime/ArrayBuffer.h

```diff
@@ -17,29 +17,33 @@ class ArrayBuffer : public Object {
 
 public:
     static ArrayBuffer* create(GlobalObject&, size_t);
-    static ArrayBuffer* create(GlobalObject&, ByteBuffer&);
     static ArrayBuffer* create(GlobalObject&, ByteBuffer*);
 
     ArrayBuffer(size_t, Object& prototype);
-    ArrayBuffer(ByteBuffer& buffer, Object& prototype);
     ArrayBuffer(ByteBuffer* buffer, Object& prototype);
     virtual ~ArrayBuffer() override;
 
     size_t byte_length() const { return buffer_impl().size(); }
     ByteBuffer& buffer() { return buffer_impl(); }
     const ByteBuffer& buffer() const { return buffer_impl(); }
 
+    Value detach_key() const { return m_detach_key; }
+    void detach_buffer() { m_buffer = Empty {}; }
+    bool is_detached() const { return m_buffer.has<Empty>(); }
+
 private:
     ByteBuffer& buffer_impl()
     {
+        VERIFY(!is_detached());
         ByteBuffer* ptr { nullptr };
-        m_buffer.visit([&](auto* pointer) { ptr = pointer; }, [&](auto& value) { ptr = &value; });
+        m_buffer.visit([&](Empty) { VERIFY_NOT_REACHED(); }, [&](auto* pointer) { ptr = pointer; }, [&](auto& value) { ptr = &value; });
         return *ptr;
     }
 
     const ByteBuffer& buffer_impl() const { return const_cast<ArrayBuffer*>(this)->buffer_impl(); }
 
-    Variant<ByteBuffer, ByteBuffer*> m_buffer;
+    Variant<Empty, ByteBuffer, ByteBuffer*> m_buffer;
+    Value m_detach_key;
```

Comment:
Since this isn't used anywhere in LibJS, could you please add a comment explaining that it shouldn't be removed? :D